### PR TITLE
Lab3 : Project

### PR DESCRIPTION
**Pull request by UsusCimex**

### CODE REVIEW

**Comment by k-stolyarov** [comment]
what task should I review ?

**Comment by UsusCimex** [comment]
> what task should I review ?

httpProxy

File: projects/httpProxy.c [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+    // Подключение к целевому серверу
+    target_sock = connect_to_target(host, PORT); // Используем PORT, определенный в начале программы
+    if (target_sock < 0) {
+        close(client_sock);
+        return NULL;
+    }
+
+    // Передача запроса на целевой сервер
+    send(target_sock, buffer, strlen(buffer), 0);
```
Comment (k-stolyarov) [inline projects/httpProxy.c]: What happen if  I try to pass binary file as HTTP request body  ?

File: projects/httpProxy.c [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    char buffer[BUFFER_SIZE];
+    int target_sock, read_bytes;
+
+    // Чтение запроса от клиента
+    read_bytes = recv(client_sock, buffer, BUFFER_SIZE - 1, 0);
+    if (read_bytes <= 0) {
+        close(client_sock);
+        return NULL;
+    }
+    buffer[read_bytes] = '\0'; // Добавляем нуль-терминатор
```
Comment (k-stolyarov) [inline projects/httpProxy.c]: what happen if request is longer than read_bytes ?

File: projects/httpProxy.c [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    int sock;
+
+    // Получение информации о хосте
+    if ((he = gethostbyname(host)) == NULL) {
+        herror("gethostbyname failed");
+        return -1;
+    }
+
+    // Создание сокета
+    if ((sock = socket(AF_INET, SOCK_STREAM, 0)) < 0) {
```
Comment (k-stolyarov) [inline projects/httpProxy.c:158]: Please list in reply all possible socket doimains, protocol and types with brief description

File: projects/httpProxy.c [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    while (1) {
+        // Принятие соединения от клиента
+        if ((client_socket = accept(server_fd, (struct sockaddr*)&address, (socklen_t*)&addr_len)) < 0) {
+            perror("accept");
+            continue;
+        }
+
+        // Создание потока для обработки соединения
+        pthread_t thread_id;
+        if (pthread_create(&thread_id, NULL, handle_connection, (void*)&client_socket) != 0) {
```
Comment (k-stolyarov) [inline projects/httpProxy.c]: what about joining threads ?

**Review by k-stolyarov — CHANGES_REQUESTED** [review]

File: projects/httpProxy.c (hunk 1 above)
Comment (UsusCimex) [reply projects/httpProxy.c]: strlen() does not work correctly with the binary format (find \0 symbol). change strlen to read_bytes. Fix it in last commit

File: projects/httpProxy.c (hunk 2 above)
Comment (UsusCimex) [reply projects/httpProxy.c]: we use TCP protocol and blocking recv so this shouldn't happen. But if the user deliberately sends the request in parts, then such a case is not processed (Is it necessary?)

File: projects/httpProxy.c (hunk 3 above)
Comment (UsusCimex) [reply projects/httpProxy.c:158]: Domains and protocols:
AF_UNIX, AF_LOCAL -> Local communication
AF_INET -> IPv4 Internet protocols
AF_INET6 -> IPv6 Internet protocols
AF_IPX -> IPX - Novell protocols
AF_NETLINK -> Kernel user interface device
AF_X25 -> ITU-T X.25 / ISO-8208 protocol
AF_AX25 -> Amateur radio AX.25 protocol
AF_ATMPVC -> Access to raw ATM PVCs
AF_APPLETALK -> Appletalk
AF_PACKET -> Low level packet interface

socket types:
SOCK_STREAM Provides sequenced, reliable, two-way, connection-based byte streams. An out-of-band data transmission mechanism may be supported.
SOCK_DGRAM Supports datagrams (connectionless, unreliable messages of a fixed maximum length).
SOCK_SEQPACKET Provides a sequenced, reliable, two-way connection-based data transmission path for datagrams of fixed maximum length; a consumer is required to read an entire packet with each input system call.
SOCK_RAW Provides raw network protocol access.
SOCK_RDM Provides a reliable datagram layer that does not guarantee ordering.
SOCK_PACKET Obsolete and should not be used in new programs.
SOCK_NONBLOCK Set the O_NONBLOCK file status flag on the new open file description.
SOCK_CLOEXEC Set the close-on-exec (FD_CLOEXEC) flag on the new file descriptor.

File: projects/httpProxy.c (hunk 4 above)
Comment (UsusCimex) [reply projects/httpProxy.c]: Can I use detach threads? If not, then I’ll think about storing and joining threads. (Simply calling Join all the time can greatly disrupt the efficiency of the program, creating an array, and if it overflows, go through the elements with a call to join - with a large flow of connections, we can hang on Join. Ideally, just implement a thread pool)

File: projects/httpProxy.c (hunk 2 above)
Comment (k-stolyarov) [reply projects/httpProxy.c]: Yes, it is required.

File: projects/httpProxy.c (hunk 4 above)
Comment (k-stolyarov) [reply projects/httpProxy.c]: Use any solution. We have limitations on amount of threads (lab 1)  and it is not desired to have the same issue here.

**Review by k-stolyarov — CHANGES_REQUESTED** [review]
fix all remainig notes from above

File: projects/httpProxy.c (hunk 2 above)
Comment (UsusCimex) [reply projects/httpProxy.c]: fixed

File: projects/httpProxy.c (hunk 4 above)
Comment (UsusCimex) [reply projects/httpProxy.c]: fixed

File: projects/httpProxy.c [hunk 5]
```diff
@@ -51,19 +52,32 @@ int main(int argc, char* argv[]) {
...
         if ((client_socket = accept(server_fd, (struct sockaddr*)&address, (socklen_t*)&addr_len)) < 0) {
             perror("accept");
             continue;
         }
 
-        // Создание потока для обработки соединения
-        pthread_t thread_id;
-        if (pthread_create(&thread_id, NULL, handle_connection, (void*)&client_socket) != 0) {
+        if (threads[current_thread] != 0) {
+            pthread_join(threads[current_thread], NULL);
```
Comment (k-stolyarov) [inline projects/httpProxy.c]: You have one long-running download thread and all remaining are for short pieces of data.  As a result you'll stuck in waiting this long thread while all remaining transfers are finished long time ago.

File: projects/httpProxy.c [hunk 6]
```diff
@@ -73,19 +87,31 @@ void* handle_connection(void* client_socket) {
...
-        close(client_sock);
-        return NULL;
+    while (1) {
+        read_bytes = recv(client_sock, buffer, BUFFER_SIZE, 0);
+        if (read_bytes <= 0) {
+            break;
+        }
+
+        data = realloc(data, data_size + read_bytes);
+        memcpy(data + data_size, buffer, read_bytes);
```
Comment (k-stolyarov) [inline projects/httpProxy.c:87]: memory leak detected.

**Review by k-stolyarov — CHANGES_REQUESTED** [review]

File: projects/httpProxy.c (hunk 5 above)
Comment (UsusCimex) [reply projects/httpProxy.c]: fixed,made threads detach

File: projects/httpProxy.c (hunk 6 above)
Comment (UsusCimex) [reply projects/httpProxy.c:87]: fixed

**Review by k-stolyarov — APPROVED** [review]
Accepted

**Pull request merged**